pysoa/common/logging.py

```diff
@@ -199,3 +202,178 @@ def _copy_and_censor_dict(cls, d):
+                try:
+                    index = formatted_message.index(record.getMessage()[:40])
+                    start_of_message, to_chunk = formatted_message[:index], formatted_message[index:]
+                except (TypeError, ValueError):
+                    # We can't locate the message in the formatted record? That's unfortunate. Let's make something up.
+                    start_of_message, to_chunk = '{} '.format(formatted_message[:30]), formatted_message[30:]
+
+                start_of_message = start_of_message.encode('utf-8')
+                to_chunk = to_chunk.encode('utf-8')
+
+                # 12 is the length of "... (cont'd)" in bytes
+                chunk_length_limit = message_length_limit - len(start_of_message) - 12
+
+                i = 1
+                parts = []
+                remaining_message = to_chunk
+                while remaining_message:
+                    message_id = b''
+                    subtractor = 0
+                    if i > 1:
+                        # If this is not the first message, we determine message # so that we can subtract that length
+                        message_id = '{}'.format(i).encode('utf-8')
+                        # 14 is the length of "(cont'd #) ..." in bytes
+                        subtractor = 14 + len(message_id)
+                    chunk, remaining_message = self._cleanly_slice_encoded_string(
+                        remaining_message,
+                        chunk_length_limit - subtractor,
+                    )
+                    if i > 1:
+                        # If this is not the first message, we prepend the chunk to indicate continuation
+                        chunk = b"(cont'd #" + message_id + b') ...' + chunk
+                    i += 1
+                    if remaining_message:
+                        # If this is not the last message, we append the chunk to indicate continuation
+                        chunk = chunk + b"... (cont'd)"
+                    parts.append(priority + prefix + start_of_message + chunk + suffix)
+
+            self._send(parts)
+        except Exception:
+            self.handleError(record)
+
+    def _send(self, parts):
+        for message in parts:
+            if self.unixsocket:
+                try:
+                    self.socket.send(message)
+                except OSError:
+                    self.socket.close()
+                    self._connect_unixsocket(self.address)
+                    self.socket.send(message)
+            elif self.socktype == socket.SOCK_DGRAM:
+                self.socket.sendto(message, self.address)
+            else:
+                self.socket.sendall(message)
+
+    @staticmethod
+    def _cleanly_slice_encoded_string(encoded_string, length_limit):
+        sliced, remaining = encoded_string[:length_limit], encoded_string[length_limit:]
+        try:
+            sliced.decode('utf-8')
+        except UnicodeDecodeError as e:
```

Comment:
OK, so I figures out why this will _always_ work correctly, but a comment in the code would be nice. It works, because the first byte of a UTF-8 character encodes how many bytes are in total in that character.

Reply:
Docstring added.